plugin/lighthouse/handler.go

```diff
@@ -0,0 +1,61 @@
+package lighthouse
+
+import (
+	"context"
+	"net"
+	"strings"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/pkg/nonwriter"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+// ServeDNS implements the plugin.Handler interface.
+func (lh *Lighthouse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+
+	a := new(dns.Msg)
+	a.SetReply(r)
+	a.Authoritative = true
+
+	services := lh.SvcsMap
+	svcName := strings.Split(state.QName(), ".")[0]
+	ip := services[svcName]
+	if ip == "" {
+		return onFailure(lh, ctx, w, r)
+	}
+	rr := new(dns.A)
+
+	if state.Family() == 1 {
+		// IPv4 query
+		rr.Hdr = dns.RR_Header{Name: state.QName(), Rrtype: dns.TypeA, Class: state.QClass()}
+		rr.A = net.ParseIP(ip).To4()
+	} else {
+		log.Debugf("IPv6 queries not supported yet")
+		return onFailure(lh, ctx, w, r)
+	}
+
+	a.Answer = []dns.RR{rr}
+
+	log.Debugf("Responding to query with '%s'", a.Answer)
+	if w.WriteMsg(a) != nil {
+		log.Debugf("Failed to writeMsg")
+		return onFailure(lh, ctx, w, r)
+	}
+
+	return dns.RcodeSuccess, nil
+}
+
+func onFailure(lh *Lighthouse, ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
```

Comment:
One source of incomprehension for me, initially, is the name of this function. I think it would be easier to understand if it was named after what it ultimately does, `nextOrFailure`.

Reply:
This is more about what `lighthouse` will do with a failure. If we later decide to have some failure scenario where we just return without checking for next we could revisit this.